textx/registration.py

```diff
@@ -0,0 +1,312 @@
+"""
+Languages and generators registration and discovery API.
+"""
+from __future__ import unicode_literals
+import fnmatch
+import pkg_resources
+
+from textx.exceptions import TextXRegistrationError
+
+
+class LanguageDesc(object):
+    """
+    A class used in language registration/discovery.
+
+    Attributes:
+        name (str): the name/ID of the language (must be unique)
+        pattern (str): filename pattern for models (e.g. "*.data")
+        description (str): A short description of the language
+        metamodel (callable): A callable that returns configured meta-model
+            or the metamodel itself (if a single specific instance is
+            desired)
+    """
+
+    def __init__(self, name, pattern=None, description='', metamodel=None):
+        self.name = name
+        self.pattern = pattern
+        self.description = description
+        self.metamodel = metamodel
+
+
+class GeneratorDesc(object):
+    """
+    A class used in generators registration/discovery.
+
+    Attributes:
+        language (str): The name/ID of the language this generator is for.
+                        If the generators is generic (applicable to any model)
+                        use "any".
+        target (str): A short name of the target stack/technology.
+        description (str): A short description of the generator.
+        generator (callable): A callable of the form:
+                              def generator(metamodel, model, output_path,
+                                            overwrite, debug, **custom_args)
+    """
+    def __init__(self, language, target, description='', generator=None):
+        self.language = language
+        self.target = target
+        self.description = description
+        self.generator = generator
+
+
+metamodels = {}
+languages = None
+generators = None
+
+
+def language_descriptions():
```

Comment:
A flag parameter like `reload_anyway` will be useful, as well for `generator_descriptions`.
I had to call `pkg_resources.WorkingSet().iter_entry_points(group='textx_languages')` in order to discover newly installed languages in already running process.

Reply:
I see now that you have already added functions to clear languages and generators, sorry.

Reply:
Yes. When the registration is cleared declaratively registered registrations gets reloaded with the next call to the API.

Reply:
I think we still need to change to:

_pkg_resources`.WorkingSet()`.iter_entry_points(group='textx_languages')_

to reinitialize _entries_ field of working set, because `pip install ...` adds a new path on `sys.path`. Also, in `generator_descriptions` function.

https://github.com/pypa/setuptools/blob/8d4e4bc93e319275458dcc7381ec83d447eca770/pkg_resources/__init__.py#L564

Reply:
You are right. Fixed. Thanks.